indico/modules/categories/operations.py

```diff
@@ -20,12 +20,12 @@
 
 
 def create_category(parent, data):
-    category = Category(parent=parent)
+    category = Category(parent=parent, title='default_title')
```

Comment:
ah right we do not allow empty titles ;)

anyway, just a thought: it'd be nicer to `category.populate_from_dict(data, keys={'title'})` instead of setting this default title. that way it's already the proper one and we never get cases where the title from data is actually missing and we are stuck with the "bad" one

Reply:
After testing it, it doesn't work because it flushes the Category before populating it, and title cannot be null

Reply:
You can use `with db.session.no_autoflush:` or put the `@no_autoflush` decorator on the function.

Reply:
can you please provide a traceback and the error message?

Reply:
Category.title is null:
```
Call stack:
  File "/Users/jbtwist/.vscode/extensions/ms-python.debugpy-2024.3.10731008-darwin-x64/bundled/libs/debugpy/_vendored/pydevd/_pydev_bundle/pydev_monkey.py", line 1118, in __call__
    ret = self.original_func(*self.args, **self.kwargs)
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/threading.py", line 930, in _bootstrap
    self._bootstrap_inner()
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/socketserver.py", line 683, in process_request_thread
    self.finish_request(request, client_address)
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/socketserver.py", line 360, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/socketserver.py", line 747, in __init__
    self.handle()
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/werkzeug/serving.py", line 390, in handle
    super().handle()
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/http/server.py", line 427, in handle
    self.handle_one_request()
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/http/server.py", line 415, in handle_one_request
    method()
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/werkzeug/serving.py", line 362, in run_wsgi
    execute(self.server.app)
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/werkzeug/serving.py", line 325, in execute
    for data in application_iter:
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/werkzeug/debug/__init__.py", line 330, in debug_application
    app_iter = self.app(environ, start_response)
  File "/Users/jbtwist/indico-core/src/indico/cli/devserver.py", line 195, in __call__
    return rv(environ, start_response)
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/sentry_sdk/integrations/flask.py", line 85, in sentry_patched_wsgi_app
    return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/sentry_sdk/integrations/wsgi.py", line 108, in __call__
    rv = self.app(
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/sentry_sdk/integrations/flask.py", line 85, in <lambda>
    return SentryWsgiMiddleware(lambda *a, **kw: old_app(self, *a, **kw))(
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/flask/app.py", line 1488, in __call__
    return self.wsgi_app(environ, start_response)
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/flask/app.py", line 1463, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/flask/app.py", line 872, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/flask/app.py", line 762, in handle_user_exception
    return self.ensure_sync(handler)(e)  # type: ignore[no-any-return]
  File "/Users/jbtwist/indico-core/src/indico/web/flask/errors.py", line 105, in handle_databaseerror
    return handle_exception(exc, _('There was a database error while processing your request.'))
  File "/Users/jbtwist/indico-core/src/indico/web/flask/errors.py", line 110, in handle_exception
    Logger.get('flask').exception(str(exc) or 'Uncaught Exception')
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/logging/__init__.py", line 1481, in exception
    self.error(msg, *args, exc_info=exc_info, **kwargs)
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/logging/__init__.py", line 1475, in error
    self._log(ERROR, msg, args, **kwargs)
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/logging/__init__.py", line 1589, in _log
    self.handle(record)
  File "/Users/jbtwist/.pyenv/versions/3.9.9/lib/python3.9/logging/__init__.py", line 1599, in handle
    self.callHandlers(record)
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/sentry_sdk/integrations/logging.py", line 99, in sentry_patched_callhandlers
    return old_callhandlers(self, record)
Message: '(psycopg2.errors.NotNullViolation) null value in column "title" of relation "categories" violates not-null constraint\nDETAIL:  Failing row contains (3, 0, f, 1, null, null, null, null, null, UTC, {"lecture": "lecture", "meeting": "standard"}, 1, {}, 0, , f, f, f, null, null, null, , 1, , 0, 2, {}).\n\n[SQL: INSERT INTO categories.categories (parent_id, is_deleted, position, visibility, icon_metadata, icon, logo_metadata, logo, timezone, default_event_themes, event_creation_mode, event_creation_notification_emails, event_message_mode, event_message, suggestions_disabled, notify_managers, show_future_months, google_wallet_mode, google_wallet_settings, is_flat_view_enabled, default_ticket_template_id, default_badge_template_id, title, description, protection_mode, no_access_contact) VALUES (%(parent_id)s, %(is_deleted)s, %(position)s, %(visibility)s, %(icon_metadata)s, %(icon)s, %(logo_metadata)s, %(logo)s, %(timezone)s, %(default_event_themes)s, %(event_creation_mode)s, %(event_creation_notification_emails)s::VARCHAR[], %(event_message_mode)s, %(event_message)s, %(suggestions_disabled)s, %(notify_managers)s, %(show_future_months)s, %(google_wallet_mode)s, %(google_wallet_settings)s, %(is_flat_view_enabled)s, %(default_ticket_template_id)s, %(default_badge_template_id)s, %(title)s, %(description)s, %(protection_mode)s, %(no_access_contact)s) RETURNING categories.categories.id]\n[parameters: {\'parent_id\': 0, \'is_deleted\': False, \'position\': 1, \'visibility\': None, \'icon_metadata\': \'null\', \'icon\': None, \'logo_metadata\': \'null\', \'logo\': None, \'timezone\': \'UTC\', \'default_event_themes\': \'{"meeting": "standard", "lecture": "lecture"}\', \'event_creation_mode\': <EventCreationMode.restricted: 1>, \'event_creation_notification_emails\': [], \'event_message_mode\': <EventMessageMode.disabled: 0>, \'event_message\': \'\', \'suggestions_disabled\': False, \'notify_managers\': False, \'show_future_months\': 0, \'google_wallet_mode\': <InheritableConfigMode.inheriting: 2>, \'google_wallet_settings\': \'{}\', \'is_flat_view_enabled\': False, \'default_ticket_template_id\': None, \'default_badge_template_id\': None, \'title\': None, \'description\': \'\', \'protection_mode\': <ProtectionMode.inheriting: 1>, \'no_access_contact\': \'\'}]\n(Background on this error at: https://sqlalche.me/e/14/gkpj)'
Arguments: ()
```

Reply:
That traceback is missing the actual source of the exception...

Reply:
```
File "/Users/jbtwist/indico-core/src/indico/modules/categories/controllers/[management.py](https://management.py/)", line 287, in _process
    new_category = create_category(self.category, [form.data](https://form.data/))
  File "/Users/jbtwist/indico-core/src/indico/modules/categories/[operations.py](https://operations.py/)", line 26, in create_category
    db.session.flush()
  File "<string>", line 2, in flush
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3449, in flush
    self._flush(objects)
  File "/Users/jbtwist/indico-core/src/.venv/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3589, in _flush
    transaction.rollback(_capture_exception=True)
  File "/Users/jbtwist/indico-core/src/indico/web/flask/[util.py](https://util.py/)", line 80, in wrapper
    return obj().process()
  File "/Users/jbtwist/indico-core/src/indico/web/[rh.py](https://rh.py/)", line 309, in process
    handle_sqlalchemy_database_error()  # this will re-raise an exception
  File "/Users/jbtwist/indico-core/src/indico/web/[rh.py](https://rh.py/)", line 299, in process
    res = self._do_process()
  File "/Users/jbtwist/indico-core/src/indico/web/[rh.py](https://rh.py/)", line 267, in _do_process
    rv = self._process()
  File "/Users/jbtwist/indico-core/src/indico/modules/categories/controllers/[management.py](https://management.py/)", line 287, in _process
    new_category = create_category(self.category, [form.data](https://form.data/))
  File "/Users/jbtwist/indico-core/src/indico/modules/categories/[operations.py](https://operations.py/)", line 26, in create_category
    db.session.flush()
```

Reply:
```patch
index d8bfb3e670..e12eb41198 100644
--- a/indico/modules/categories/operations.py
+++ b/indico/modules/categories/operations.py
@@ -21,11 +21,13 @@ from indico.util.string import crc32

 def create_category(parent, data):
     category = Category(parent=parent)
+    category.populate_from_dict(data, keys=('title',))
     data.setdefault('default_event_themes', parent.default_event_themes)
     data.setdefault('timezone', parent.timezone)
-    category.populate_from_dict(data)
     db.session.add(category)
     db.session.flush()
+    category.populate_from_dict(data)
+    db.session.flush()
     signals.category.created.send(category)
     logger.info('Category %s created by %s', category, session.user)
     sep = ' \N{RIGHT-POINTING DOUBLE ANGLE QUOTATION MARK} '
```

this fixes the problem for me...